### ISSUE TITLE
Client ConnectAsync does not provide sensible exception when mqtt broker is offline

### ISSUE THREAD

**Issue by dpsenner** [chkr1011/MQTTnet#249]
# Expected behavior

When the mqtt broker is offline and ConnectAsync is invoked, an exception should be thrown that makes it clear that it cannot connect to the mqtt broker because it is unreachable.

# Actual behavior

Interestingly, we observe a wide variety of exceptions and most of them are unrelated to the actual problem. Here's a list of the exceptions that we observed so far, the first being the most sensible of the following list:

```text
MQTTnet.Exceptions.MqttCommunicationTimedOutException: Exception of type 'MQTTnet.Exceptions.MqttCommunicationTimedOutException' was thrown.
   at MQTTnet.Internal.TaskExtensions.<TimeoutAfter>d__1`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttPacketDispatcher.<WaitForPacketAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<SendAndReceiveAsync>d__41`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<AuthenticateAsync>d__32.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Subscriber.SubscriberApplicationContext.<ConnectAsync>d__23.MoveNext() in C:\Workspace\Development\studies\evaluate-mqtt\src\clients\MQTTnet\DotNETCore\Subscriber\SubscriberApplicationContext.cs:line 87
```

```text
System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<StartReceivingPacketsAsync>d__45.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Subscriber.SubscriberApplicationContext.<ConnectAsync>d__23.MoveNext() in C:\Workspace\Development\studies\evaluate-mqtt\src\clients\MQTTnet\DotNETCore\Subscriber\SubscriberApplicationContext.cs:line 87
```

```text
MQTTnet.Exceptions.MqttCommunicationException: The semaphore has been disposed. ---> System.ObjectDisposedException: The semaphore has been disposed.
   at System.Threading.SemaphoreSlim.CheckDispose()
   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
   at MQTTnet.Adapter.MqttChannelAdapter.<>c__DisplayClass12_0.<<SendPacketsAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Adapter.MqttChannelAdapter.<ExecuteAndWrapExceptionAsync>d__15.MoveNext()
   --- End of inner exception stack trace ---
   at MQTTnet.Adapter.MqttChannelAdapter.<ExecuteAndWrapExceptionAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<SendAndReceiveAsync>d__41`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<AuthenticateAsync>d__32.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Subscriber.SubscriberApplicationContext.<ConnectAsync>d__23.MoveNext() in C:\Workspace\Development\studies\evaluate-mqtt\src\clients\MQTTnet\DotNETCore\Subscriber\SubscriberApplicationContext.cs:line 87
```

```text
System.OperationCanceledException: The operation was canceled.
   at MQTTnet.Adapter.MqttChannelAdapter.<ExecuteAndWrapExceptionAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<SendAndReceiveAsync>d__41`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<AuthenticateAsync>d__32.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Subscriber.SubscriberApplicationContext.<ConnectAsync>d__23.MoveNext() in C:\Workspace\Development\studies\evaluate-mqtt\src\clients\MQTTnet\DotNETCore\Subscriber\SubscriberApplicationContext.cs:line 87
```

```text
System.NullReferenceException: Object reference not set to an instance of an object.
   at MQTTnet.Client.MqttClient.SendAsync(MqttBasePacket[] packets)
   at MQTTnet.Client.MqttClient.<SendAndReceiveAsync>d__41`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<AuthenticateAsync>d__32.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at MQTTnet.Client.MqttClient.<ConnectAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Subscriber.SubscriberApplicationContext.<ConnectAsync>d__23.MoveNext() in C:\Workspace\Development\studies\evaluate-mqtt\src\clients\MQTTnet\DotNETCore\Subscriber\SubscriberApplicationContext.cs:line 87
```

**Comment by chkr1011** [comment]
Hi, which version do you use? I fixed this or a similar issue in 2.7.5. 
Best regards 
Christian

**Comment by AtosNicoS** [comment]
I am also seeing this problem with 2.7.5. I am connecting to `broker.hivemq.com` with most settings set to default. I see the 2nd exception you posted within the Disconnect event. `e.Exception` contains `System.Threading.Tasks.TaskCanceledException: 'A task was canceled.'`.

I can reproduce this every time I am running my program for about 10-20 seconds. My client reconnects and will disconnect after a few seconds again. Today I observed that this behavior stops after some time, but I am not 100% sure about this behavior.

This is how I created the MQTTClient:
```csharp
public Actor(string topic)
{
    Topic = topic;

    // Last will & testament
    lwt = new MqttApplicationMessage()
    {
        Topic = Topic + "status",
        Payload = Encoding.UTF8.GetBytes("disconnect"),
        QualityOfServiceLevel = MqttQualityOfServiceLevel.AtLeastOnce,
        Retain = true
    };

    // MQTT client options
    options = new MqttClientOptions()
    {
        ClientId = ClientId, // Guid.NewGuid().ToString();
        WillMessage = lwt,
        ChannelOptions = new MqttClientTcpOptions()
        {
            Server = Broker // "broker.hivemq.com"
        }
    };

    // Create a new MQTT client.
    var factory = new MqttFactory();
    _mqttClient = factory.CreateMqttClient();

    // Message received callback
    _mqttClient.ApplicationMessageReceived += MqttClientOnApplicationMessageReceived;
    _mqttClient.Connected += MqttClientOnConnected;
    _mqttClient.Disconnected += MqttClientOnDisconnected;
}
```

**Comment by chkr1011** [comment]
Please try again with the latest dev branch or a preview version of 3.0.0 and reopen this ticket if it still fails.

**Comment by arnvidhr** [comment]
I am experiencing similar issue, when trying to log in. I use the MQTT mosquitto broker, almost with default settings, but only after enabling the log & anonymous access. The goal is only to test everything locally. Emulating several MQTT devices for further testing and the application just stopping execution during attempt to connect, even the exception handler doesn't fire, until I found how to add your CancellationTokenSource and implemented some kind of Logger from your samples, after which I saw the error message just like here, but without more detailed explanation, why that happens..

Using recent NuGet version 4.3.3.952... What's interesting is that it worked a little at the very beginning... But it doesn't work anymore, whatever I try, while I didn't changed anything about connection. I thought maybe the problem is in mosquitto broker, but it do not register even attempt to connect. Everything on localhost.

MyLogger.cs::Publish(45) MqttClientTrying to connect with server '127.0.0.1:1883'
MyLogger.cs::Publish(45) MqttClient:Error while connecting with server, The operation was canceled.
MyLogger.cs::Publish(45) MqttClientDisconnecting [Timeout=00:01:40]
MyLogger.cs::Publish(45) MqttClient:Disconnected from adapter.
MyLogger.cs::Publish(45) MqttClient:Disconnected.
